
## FIU Genomics & Bioinformatics Analysis CURE BSC3990L Genomics ###

***Genome assembly tutorial***

 ### Login to the FIU HPC And Linux refresher
We will be logging onto the HPCC and doing some preliminary navigation. You will use your panther credentials (username and password) to connect.

Login at the FIU Panther cluster Web portal from a browser: https://hpcgui.fiu.edu Note: Mozilla Firefox has not been working for me lately and I have to connect via Safari.

Once you are on, go to the top where it says 'Clusters' and select 'Panther Shell Access.' It will open a terminal and ask you for your panther credentials again.

To see what directory (folder) you are currently in type:

    pwd

To see what files and directories (folders) are in your current directory (folder) type:

    ls

If you want to move up 1 directory type:

    cd ..

To go back to your previous directory type:

    cd -


## What is assembly, a quick review

There are two types of assembly, reference based (generally more referred to as alignment), and de novo, meaning "from the beginning." When choosing between the two, it is essential to keep in mind the purpose of your experiment as well as data available. Alignment is used when there is a reference genome available and your goal is more to understand variation in populations. Alignment has less strigent requirements, being possible with 20-30x depth of coverage and shorter reads. de novo assembly is what we will focus on today. It is used when a species has never been sequenced before. For de novo assembly it is generally recommended to have 60x depth of coverage and some type of long read sequencing. 

Genome assembly is like a puzzel. The ideal assembly is exactly the number of chromosomes a species has. For _caenorhabditis_ nematodes, their genome contains 6 chromosomes, 5 autosomes and an X (they have an XX/XO sex determination system, instead of XX/XY like humans). Thus, the ideal assembly for these nematodes would be 6 pieces. But puzzels can be tricky sometimes, like if the pieces are very small (short read sequencing), if pieces are missing (lack of coverage), or if pieces are identical (repeat regions). You will see in this tutorial that genome assembly varies wildly with the amount of data used and your choice of assembly software.

## Obtain the data

Although each of you have your own data, we will use some test data for this assembly exercise so we are all on the same page. Some of these programs can also take up to a day to run depending on the amount of data you have. On the other hand, following this tutorial with the test data should only take an hour or two depending on your pace.

    cp /home/data/jfierst/JU4121/cat_fastq_pass_JU4121_noClassData/JU4121_classPractice_ont.fastq ~/.

Check file size to see how much data you have. Type:

    ls -lath JU4121_classPractice_ont.fastq

Look at the file. Type:

    head JU4121_classPractice_ont.fastq

Remember the format of a fastq? A fastq has 4 lines. The first begins with an @ sign. This is the sequence identifier line. In our data this line also contains useful information about the nanopore run, such as sample_id (except ours is blank because we did not give it a sample id), the date and time of the experiment, and the basecalling model used. The second line is the sequence itself (ATCG). The third is usually just a + sign, but sometimes repeats the identifier line. Finally, the fourth is quality information and should look like random symbols ("#$$''+''&$&'). For example, a $ sign is a Q score of 3, meaning that there is only a 50% chance that that base was called correctly. 

## Quality Check the Reads

There are several ways to look at the raw data. Most programs are meant for Illumina reads, but we can co-opt them for our nanopore data. Below are three ways to check your reads.

<details>
<summary><b>Depth of Coverage</b></summary>

We are going to use the awk language to process our fastq file and give us the total amount of base pairs in the fastq file.

    awk '{if(NR%4==2) total+=length($0)} END {print total}' JU4121_classPractice_ont.fastq > total_amount_of_base_pairs.txt

Once you see the prompt ($) then your command finished. 

This command is saying that if the current line number equals the second of every group of four lines, then count the characters and add the count to total. At the END, print the total and store it in total_amount_of_base_pairs.txt 

Make sure total_amount_of_base_pairs.txt was created and is in the current directory by typing:
    
    ls
    
Now view the contents in the total_amount_of_base_pairs.txt, type:

    head total_amount_of_base_pairs.txt
 
It's time to calculate the sequencing depth, use a calculator.

 Sequencing depth of coverage formula: Total base pairs sequenced / Genome sizeâ€‹	( In our case the genome size will be 100,000,000 == 100Mega base pairs)

 What was your sequencing depth? (please write this down)

  In case your curious this is the full expanded form of Sequencing depth of coverage formula: 

  Sequencing depth of coverage = (Read length * Number of reads) / Genome size

</details>


<details>
<summary><b>Fastqc</b></summary>

    #!/bin/bash

    #SBATCH --account acc_jfierst_classroom
    #SBATCH --partition highmem1
    #SBATCH --qos highmem1
    #SBATCH -n 8
    #SBATCH --output=output_fastqc.log

    module load fastqc-0.11.7-gcc-4.8.5-72rtvom

    mkdir -p fastqc_out

    fastqc JU4121_classPractice_ont.fastq -t 8 -o fastqc_out

fastqc finishes in under a minute.

</details>

<details>
<summary><b>kmer graphs with genomeScope</b></summary>
 
</details>
