
# FIU Genomics & Bioinformatics Analysis CURE BSC3990L Genomics ###

***Genome assembly tutorial***

 ## Login to the FIU HPC And Linux refresher
We will be logging onto the HPCC and doing some preliminary navigation. You will use your panther credentials (username and password) to connect.

Login at the FIU Panther cluster Web portal from a browser: https://hpcgui.fiu.edu Note: Mozilla Firefox has not been working for me lately and I have to connect via Safari.

Once you are on, go to the top where it says 'Clusters' and select 'Panther Shell Access.' It will open a terminal and ask you for your panther credentials again.

To see what directory (folder) you are currently in type:

    $ pwd

To see what files and directories (folders) are in your current directory (folder) type:

    $ ls

If you want to move up 1 directory type:

    $ cd ..

To go back to your previous directory type:

    $ cd -

## Find Sequencing depth of coverage of raw sequences
Brief is it neccessay?

We are going to use the seqtk program to process our fastq file and give us the total amount of base pairs in the fastaq file.

First we need to see what is the latest version of seqtk is on the FIU HPC,
so Type:

    module avail

This command shows us all the programs on the FIU HPC press the space bar to go down the list and search for the lastest seqtk version. Once you have found the latest seqtk version type:

    module load seqtk (The version you found, please don't type the parentheses)

If you don't see any output after loading seqtk that means it worked! Now we need to calculate the total number of bases in the fastq file and store the results in txt file. So type: (replace: your fastq file with your actually fastaq file)

    seqtk seq -A your fastq file  | awk '{if(NR%2==0) total+=length($0)} END {print total}' > total_amount_of_base_pairs.txt

This may take awhile (like 7 mins), once you see the prompt ($) then your command finished. (Read length * Number of reads = Total amount of Base pairs)

You should have a text file that contains the total amount of base pairs in your fastq file called total_amount_of_base_pairs.txt. Again type ls to view the files and directories in your current directory:
    
    $ ls
Now view the contents in the total_amount_of_base_pairs.txt, type:

    $ head -n 3 total_amount_of_base_pairs.txt
 
It's time to calculate the sequencing depth, use a calculator.


 Sequencing depth of coverage formula: Total base pairs sequenced / Genome sizeâ€‹	( In our case the genome size will be 100,000,000 == 100Mega base pairs)


 What was your sequencing depth? (please write this down)

  In case your curious this is the full expanded form of Sequencing depth of coverage formula: 

  Sequencing depth of coverage = (Read length * Number of reads) / Genome size

## Genome assembly with NextDenovo

    #Create the input file
    ls SRR16242712.fastq > input.fofn

    #Create the configuration file for assembly
    vi run.cfg

Press [i] to enter insert mode and copy and paste the below section (this was obtained by going to nextDenovo documentation and copying the run.cfg file).

    [General]
    job_type = local
    job_prefix = nextDenovo
    task = all
    rewrite = yes
    deltmp = yes
    parallel_jobs = 20
    input_type = raw
    read_type = ont # Oxford Nanpore data
    input_fofn = input.fofn
    workdir = worm_data

    [correct_option]
    read_cutoff = 1k
    genome_size = 100M # Estimated genome size    
    sort_options = -m 20g -t 15
    minimap2_options_raw = -t 8
    pa_correction = 3
    correction_options = -p 15

    [assemble_option]
    minimap2_options_cns = -t 8
    nextgraph_options = -a 1
    
Save by pressing [esc], type ":w".
To exit type: ":q" and press [enter]

    #Create the script to run nextDenovo and create an assembled genome
    vi assemble.sh
Press [i] for insert mode and copy the below script

    #!/bin/bash

    #SBATCH --account iacc_jfierst
    #SBATCH --qos=highmem1
    #SBATCH --partition=highmem1
    #SBATCH --output=out_%assemble.log
    #SBATCH --mail-user=username@email.com 	#use your own email instead
    #SBATCH --mail-type=ALL

    module load nextDenovo-2.5.0

    nextDenovo run.cfg
Save by pressing [esc], type ":w". To exit type: ":q" and press [enter]

Run the script with:

    sbatch assemble.sh
To see if your job is running type the following command:

    squeue --me
The final assembly result is at 03.ctg_graph/nd.asm.fasta

When the job is over you congrats you have assembled your first genome!! But wait there's more now its time to see the quality your genome assembly.

## Quality Check
Assembly quality has various measures. Things like N50, contig number, assembly size, k-mer counting, and gene presence/absence can all be indications of how good an assembly may be. It is a good idea to try multiple assembly methods and use these metrics to compare them. The "best" assembly is usually the most complete and contiguous. QUAST is particularly nice for comparing multiple assemblies at once.

First find the assembly file store the location of the file in txt.file
    
    $ cd worm_data
    # Go to the directory that contains your genome assembly file
    $ cd 03.ctg_graph/
    # Use ls command to locate your assembly file
    $ ls
    # Your assembly file should be nd.asm.fasta

QUAST calculates assembly metrics like N50, contig #, and assembly size. If given a reference, it can do more, like tell you about misassemblies, however, if the assembly is de novo then you do not have a reference.

    #!/bin/bash
    #SBATCH ---account iacc_jfierst
    #SBATCH --qos=highmem1
    #SBATCH --partition=highmem1
    #SBATCH --output=out_%quast.log
    #SBATCH --mail-user=username@email.com   # Please use your own email
    #SBATCH --mail-type=ALL

    module load quast-5.2.0  #may need to load before running script

    quast.py -t 4 --eukaryote --plots-format pdf /your/path/to/01_rundir/genome_assembly_file.fasta -o ./nematode_quast/
    
