
## FIU Genomics & Bioinformatics Analysis CURE BSC3990L Genomics ###

***Genome assembly tutorial***

 ### Login to the FIU HPC And Linux refresher
We will be logging onto the HPCC and doing some preliminary navigation. You will use your panther credentials (username and password) to connect.

Login at the FIU Panther cluster Web portal from a browser: https://hpcgui.fiu.edu Note: Mozilla Firefox has not been working for me lately and I have to connect via Safari.

Once you are on, go to the top where it says 'Clusters' and select 'Panther Shell Access.' It will open a terminal and ask you for your panther credentials again.

To see what directory (folder) you are currently in type:

    pwd

To see what files and directories (folders) are in your current directory (folder) type:

    ls

If you want to move up 1 directory type:

    cd ..

To go back to your previous directory type:

    cd -


## Obtain the data

Although each of you have your own data, we will use some test data for this assembly exercise so we are all on the same page. Some of these programs can also take up to a day to run depending on the amount of data you have. On the other hand, following this tutorial with the test data should only take an hour or two depending on your pace.

    cp /home/data/jfierst/JU4121/cat_fastq_pass_JU4121_noClassData/JU4121_classPractice_ont.fastq ~/.

Check file size to see how much data you have. Type:

    ls -lath JU4121_classPractice_ont.fastq

Look at the file. Type:

    head JU4121_classPractice_ont.fastq

Remember the format of a fastq? A fastq has 4 lines. The first begins with an @ sign. This is the sequence identifier line. In our data this line also contains useful information about the nanopore run, such as sample_id (except ours is blank because we did not give it a sample id), the date and time of the experiment, and the basecalling model used. The second line is the sequence itself (ATCG). The third is usually just a + sign, but sometimes repeats the identifier line. Finally, the fourth is quality information and should look like random symbols ("#$$''+''&$&'). For example, a $ sign is a Q score of 3, meaning that there is only a 50% chance that that base was called correctly. 

## Quality Check the Reads

There are several ways to look at the raw data. Most programs are meant for Illumina reads, but we can co-opt them for our nanopore data. Below are three ways to check your reads.

<details>
<summary><b>Depth of Coverage with seqtk</b></summary>

We are going to use the seqtk program to process our fastq file and give us the total amount of base pairs in the fastq file.

First we need to see what is the latest version of seqtk is on the FIU HPC,
so Type:

    module avail

This command shows us all the programs on the FIU HPC. You can press the space bar to go down the list and search for the lastest seqtk version. However, there are many modules and this can be overwhelming. If you are searching for a certain module, you can also search by name. To search for seqtk specifically type: 

    module avail seqtk

Once you have found the seqtk module type:

    module load seqtk-1.3 #The version you found

To know if it worked, type:
   
    module list
    
If it worked, you should see seqtk listed in your modules. Now we need to calculate the total number of bases in the fastq file and store the results in txt file. So type: (replace: your fastq file with your actually fastaq file)

    seqtk seq -A JU4121_classPractice_ont.fastq  | awk '{if(NR%2==0) total+=length($0)} END {print total}' > total_amount_of_base_pairs.txt

Once you see the prompt ($) then your command finished. (Read length * Number of reads = Total amount of Base pairs)

You should have a text file that contains the total amount of base pairs in your fastq file called total_amount_of_base_pairs.txt. Again type ls to view the files and directories in your current directory:
    
    $ ls
Now view the contents in the total_amount_of_base_pairs.txt, type:

    $ head -n 3 total_amount_of_base_pairs.txt
 
It's time to calculate the sequencing depth, use a calculator.


 Sequencing depth of coverage formula: Total base pairs sequenced / Genome sizeâ€‹	( In our case the genome size will be 100,000,000 == 100Mega base pairs)


 What was your sequencing depth? (please write this down)

  In case your curious this is the full expanded form of Sequencing depth of coverage formula: 

  Sequencing depth of coverage = (Read length * Number of reads) / Genome size

</details>


<details>
<summary><b>Fastqc</b></summary>

    #!/bin/bash

    #SBATCH --account acc_jfierst_classroom
    #SBATCH --partition highmem1
    #SBATCH --qos highmem1
    #SBATCH -n 8
    #SBATCH --output=output_fastqc.log

    module load fastqc-0.11.7-gcc-4.8.5-72rtvom

    mkdir -p fastqc_out

    fastqc JU4121_classPractice_ont.fastq -t 8 -o fastqc_out

fastqc finishes in under a minute.

</details>

<details>
<summary><b>kmer graphs with genomeScope</b></summary>
 
</details>
