
## FIU Genomics & Bioinformatics Analysis CURE BSC3990L Genomics ###

***Genome assembly tutorial***

 ### Login to the FIU HPC And Linux refresher
We will be logging onto the HPCC and doing some preliminary navigation. You will use your panther credentials (username and password) to connect.

Login at the FIU Panther cluster Web portal from a browser: https://hpcgui.fiu.edu Note: Mozilla Firefox has not been working for me lately and I have to connect via Safari.

Once you are on, go to the top where it says 'Clusters' and select 'Panther Shell Access.' It will open a terminal and ask you for your panther credentials again.

To see what directory (folder) you are currently in type:

    $ pwd

To see what files and directories (folders) are in your current directory (folder) type:

    $ ls

If you want to move up 1 directory type:

    $ cd ..

To go back to your previous directory type:

    $ cd -

## Find Sequencing depth of coverage of raw sequences


We are going to use the seqtk program to process our fastq file and give us the total amount of base pairs in the fastaq file.

First we need to see what is the latest version of seqtk is on the FIU HPC,
so Type:

    module avail

This command shows us all the programs on the FIU HPC press the space bar to go down the list and search for the lastest seqtk version. Once you have found the latest seqtk version type:

    module load seqtk (The version you found, please don't type the parentheses)

If you don't see any output after loading seqtk that means it worked! Now we need to calculate the total number of bases in the fastq file and store the results in txt file. So type: (replace: your fastq file with your actually fastaq file)

    seqtk seq -A your fastq file  | awk '{if(NR%2==0) total+=length($0)} END {print total}' > total_amount_of_base_pairs.txt

This may take awhile (like 7 mins), once you see the prompt ($) then your command finished. (Read length * Number of reads = Total amount of Base pairs)

You should have a text file that contains the total amount of base pairs in your fastq file called total_amount_of_base_pairs.txt. Again type ls to view the files and directories in your current directory:
    
    $ ls
Now view the contents in the total_amount_of_base_pairs.txt, type:

    $ head -n 3 total_amount_of_base_pairs.txt
 
It's time to calculate the sequencing depth, use a calculator.


 Sequencing depth of coverage formula: Total base pairs sequenced / Genome sizeâ€‹	( In our case the genome size will be 100,000,000 == 100Mega base pairs)


 What was your sequencing depth? (please write this down)

  In case your curious this is the full expanded form of Sequencing depth of coverage formula: 

  Sequencing depth of coverage = (Read length * Number of reads) / Genome size

