
## FIU Genomics & Bioinformatics Analysis CURE BSC3990L Genomics ###

***Genome assembly tutorial***

 ### Login to the FIU HPC And Linux refresher
We will be logging onto the HPCC and doing some preliminary navigation. You will use your panther credentials (username and password) to connect.

Login at the FIU Panther cluster Web portal from a browser: https://hpcgui.fiu.edu Note: Mozilla Firefox has not been working for me lately and I have to connect via Safari.

Once you are on, go to the top where it says 'Clusters' and select 'Panther Shell Access.' It will open a terminal and ask you for your panther credentials again.

To see what directory (folder) you are currently in type:

    pwd

To see what files and directories (folders) are in your current directory (folder) type:

    ls

If you want to move up 1 directory type:

    cd ..

To go back to your previous directory type:

    cd -
If you want to go to your home directory type:

    cd ~

## What is assembly? A quick review

There are two types of assembly, reference based (generally more referred to as alignment), and de novo, meaning "from the beginning." When choosing between the two, it is essential to keep in mind the purpose of your experiment as well as data available. Alignment is used when there is a reference genome available and your goal is more to understand variation in populations. Alignment has less strigent requirements, being possible with 20-30x depth of coverage and shorter reads. De novo assembly is used when a species has never been sequenced before. For de novo assembly it is generally recommended to have 60x depth of coverage and some type of long read sequencing. De novo assembly is what we will focus on today and it will simply be called assembly from here on. We will also be using e.coli since some of these programs can take more hours than we have in class to run and since this is just a small tutorial.

Genome assembly is like a puzzle. The ideal assembly is exactly the number of chromosomes a species has. For E.coli (and many other bacteria), they have one circular genome approximately 5Mb in length, and possibly a few plasmids. Thus, the ideal assembly would be 1 large contig and maybe a few smaller pieces. But puzzles can be tricky sometimes, like if the puzzle has no border (not knowing the plasmid number for example), like if the pieces are very small (short read sequencing), if pieces are missing (lack of coverage), or if pieces are identical (repeat regions). You will see in this tutorial that genome assembly varies wildly with the amount of data used and your choice of assembly software.

## Obtain the data

Although each of you have your own data, we will use some test data for this assembly exercise so we are all on the same page. Some of these programs can also take up to a day to run depending on the amount of data you have. On the other hand, following this tutorial with the test data should only take an hour or two depending on your pace.

    cp /home/data/jfierst/nanopore_class_2025/assemblyPractice/ecoli.fastq ~/.

Check file size to see how much data you have. Type:

    ls -lath ecoli.fastq

Look at the file. Type:

    head -5 ecoli.fastq

Remember the format of a fastq? A fastq has 4 lines. The first begins with an @ sign. This is the sequence identifier line. In our data this line also contains useful information about the nanopore run, such as sample_id, the date and time of the experiment, and the basecalling model used. The second line is the sequence itself (ATCG). The third is usually just a + sign, but sometimes repeats the identifier line. Finally, the fourth is quality information and should look like random symbols ("#$$''+''&$&'). Each symbol is code for a quality score. For example, a $ sign is a Q score of 3, meaning that there is only a 50% chance that that base was called correctly. 

## Quality Check the Reads

There are several ways to look at the raw data. We will look at coverage and quality.

<details>
<summary><b>Depth of Coverage</b></summary>

We are going to use the awk language to process our fastq file and give us the total amount of base pairs in the fastq file.

    awk '{if(NR%4==2) total+=length($0)} END {print total}' ecoli.fastq > total_amount_of_base_pairs.txt

Once you see the prompt ($) then your command finished. 

This command is saying that if the current line number equals the second of every group of four lines, then count the characters and add the count to total. At the END, print the total and store it in total_amount_of_base_pairs.txt 

Make sure total_amount_of_base_pairs.txt was created and is in the current directory by typing:
    
    ls
    
Now view the contents in the total_amount_of_base_pairs.txt, type:

    head total_amount_of_base_pairs.txt
 
It's time to calculate the sequencing depth. The equation is: Sequencing depth of coverage formula= 
Total base pairs sequenced / Genome size​	( In our case the genome size will be 5,000,000 == 5Mb)

You may use a calculator. Alternatively, you could use awk in the command line for simple math: awk 'BEGIN { print total_base_pairs / 100000000 }' 
      
      # Replace total_base_pairs with the number stored in total_amount_of_base_pairs.txt
      
      Ex: awk 'BEGIN { print 20 / 100000000 }'

 What was your sequencing depth? (please write this down)

  In case your curious this is the full expanded form of Sequencing depth of coverage formula: 

  Sequencing depth of coverage = (Read length * Number of reads) / Genome size

</details>


<details>
<summary><b>Fastqc</b></summary>

Use vi to create a new file

    vi fastqc.sh

You should see the vi window with the ~ symbols. Type ‘i’ to get into input or insertion mode and then type the following into your file:
    
    #!/bin/bash

    #SBATCH --account acc_jfierst_classroom
    #SBATCH --partition highmem1
    #SBATCH --qos highmem1
    #SBATCH -n 8
    #SBATCH --output=output_fastqc.log
    #SBATCH --mail-user=username@email.com 	#use your own email instead
    #SBATCH --mail-type=ALL

    module load fastqc-0.11.7-gcc-4.8.5-72rtvom #code is loading the software needed

    mkdir -p fastqc_out   #code is saying if the directory doesn't already exist, make it

    fastqc ecoli.fastq -t 8 -o fastqc_out  #code is running the software on the input file with 8 threads and telling the output to be put in fastqc_out

Hit the escape button (‘esc’) to get out of insertion mode and type ‘:wq’ (without the single quotes) to save the file and exit vi. 

Submit the script:

    sbatch fastqc.sh

fastqc finishes in about 3 minutes. The output is an html file in the directory fastqc_out. To view this, you need to download the file and open it in your preferred browser. To download a file from the HPC to your local machine, 

Go to the HPC dashboard tab that is currently open on your laptop. 

Click on the Files button.

Next, click on the Home Directory button.

Navigate to the file and check the white box next to it.

Finally, click the Download button in the top-right corner to download the file.

Your fastqc graphs should look something like:![screenshot](https://github.com/FierstLab/Bootcamp/blob/main/pictures/fastqc.png)

What are your shortest and longest reads?

What is the average quality score?

Notice the first few bases are really poor quality (I might have forgotten to trim the barcodes). 

</details>

## Genome Assembly

Below are just 2 of many softwares used to assemble a genome. 

<details>
<summary><b>NextDenovo</b></summary>

NextDenovo is available on the HPC already. Try typing: module avail next

You don't need to load the module now because we will do that in the job script. 

    #Create the input file
    ls ecoli.fastq > input.fofn

    #Create the configuration file for assembly
    vi run.cfg

Press [i] to enter insert mode and copy and paste the below section (this was obtained by going to nextDenovo documentation and copying the run.cfg file).

    [General]
    job_type = local
    job_prefix = nextDenovo
    task = all
    rewrite = yes
    deltmp = yes
    parallel_jobs = 20
    input_type = raw
    read_type = ont # Oxford Nanpore data
    input_fofn = input.fofn
    workdir = nextdenovo_out

    [correct_option]
    read_cutoff = 1k
    genome_size = 5M # Estimated genome size    
    sort_options = -m 20g -t 15
    minimap2_options_raw = -t 8
    pa_correction = 3
    correction_options = -p 15

    [assemble_option]
    minimap2_options_cns = -t 8
    nextgraph_options = -a 1
    
Save by pressing [esc], type ":w".
To exit type: ":q" and press [enter]

    #Create the script to run nextDenovo and create an assembled genome
    vi assemble.sh
    
Press [i] for insert mode and copy and paste the below script

    #!/bin/bash

    #SBATCH --account iacc_jfierst
    #SBATCH --qos=highmem1
    #SBATCH --partition=highmem1
    #SBATCH --output=out_nextdenovo.log
    #SBATCH -n 20
    #SBATCH --mail-user=username@email.com 	#use your own email instead
    #SBATCH --mail-type=ALL

    module load nextDenovo-2.5.0

    nextDenovo run.cfg

Hit the escape button (‘esc’) to get out of insertion mode and type ‘:wq’ (without the single quotes) to save the file and exit vi. 
    
Run the script with:

     sbatch assemble.sh
    
To see if your job is running type the following command:

     squeue --me

Your assembly will take about 15 minutes to complete. You can try flye or subsetting your data while you wait, but make sure you come back once it has completed. 

When finished, your assembly will be in the nextdenovo_out/03.ctg_graph directory. cd into the directory and ls. You should see nd.asm.fasta. Type: ls -lath. What is the size of the file? 

Look at the file with: head nd.asm.fasta. You should notice a difference between a fasta and fastq. Unlike a fastq which has 4 lines (id, seq, +, quality), a fasta only has 2 lines. The first line is >contig_name, and the second is the sequence (ATCG). Type: 

grep -c ">" nd.asm.fasta 

How many contigs are in your assembly?

Finally, type: more nd.asm.fasta.stat
You should see some assembly statistics including total assembly length, number of contigs (fragments), contig N50, largest contig, smallest contig, and the mean depth of coverage. 

</details>


<details>
<summary><b>Flye</b></summary>

For Flye, first we must download the flye software because it is not readily available on the HPC. If you don't believe me go ahead and try module avail flye. To install flye, we will use a version of conda called mamba. Type:

    module load mamba/23.1.0-4

Now that we have mamba loaded, we need to create an environment for flye. Type:

    mamba create -n flye

Once the environment is created, activate it with:

    source activate flye

You will know it worked when (flye) appears before your username $. With flye activated, download the software with:
    
    # After typing the command below this step will take a ~ 11 mins to complete
    # Ignore the first four ERRORs you see
    mamba install bioconda::flye

Now lets make a script with vi. 

    vi flye.sh

Copy and paste the following script. Make sure to read over it and change any relavent information like your email. 

    #!/bin/bash

    #SBATCH --account iacc_jfierst
    #SBATCH --partition highmem1
    #SBATCH --qos highmem1
    #SBATCH -n 12
    #SBATCH --output=output_flye.log
    #SBATCH --mail-user=username@fiu.edu    #use your own email instead
    #SBATCH --mail-type=ALL

    module load mamba/23.1.0-4   #load the softwares
    source activate flye

    flye --nano-hq ecoli.fastq --genome-size 5m --out-dir flye_out --threads 12   #run flye, specifying high quality nanopore data, the input file, the estimated genome size assumed from e.coli, the output directory, and the number of threads the computer will use.

Hit the escape button (‘esc’) to get out of insertion mode and type ‘:wq’ (without the single quotes) to save the file and exit vi. 

Submit the script:

    sbatch flye.sh

See the status of your job:

    squeue --me
    
Your assembly will take about 15 minutes to complete. To check on the status of the job type: squeue --me. You can move forward with nextdenovo while you wait, but make sure you come back once it has completed. 

When finished, your assembly will be in the flye_out directory. cd into the directory and ls. You should see assembly.fasta. Type: ls -lath. What is the size of the file? 

Look at the file with: head assembly.fasta. You should notice a difference between a fasta and fastq. Unlike a fastq which has 4 lines (id, seq, +, quality), a fasta only has 2 lines. The first line is >contig_name, and the second is the sequence (ATCG). Type: 

    grep -c ">" assembly.fasta 

How many contigs are in your assembly?

Finally, type: tail flye.log
You should see some assembly statistics including total assembly length, number of contigs (fragments), contig N50, largest contig, and the mean depth of coverage. 

</details>


When the job is over you congrats you have assembled your first de novo genome(s)!! :) But wait there's more! Now its time to see the quality your genome assembly.

## Quality Check
Assembly quality has various measures. Things like N50, contig number, assembly size, kmers, and gene presence/absence can all be indications of how good an assembly may be. It is a good idea to try multiple assembly methods and use these metrics to compare them. The "best" assembly is usually the most complete and contiguous. QUAST and BUSCO are some of the most common programs used to evaluate assemblies. 

<details>
 <summary><b>QUAST</b></summary>
 
First find the assembly file and store the file path a in txt file
    
    # Locate your assembly file. Your assembly file should be nd.asm.fasta (nextdenovo) or assembly.fasta (flye)
    # Save the path to this file in a txt file to use for your quast script.
     pwd > assembly_file_path.txt
    # View the contents of the file
     head -n 3 assembly_file_path.txt
    ## High light the output with your cursor (mouse) and then press ctrl + c (for windows) , command + c (for macs) (This is the keyboard shortcut to copy txt, 
    we will paste the file path in the quast script later)

QUAST calculates assembly metrics like N50, contig #, and assembly size. If given a reference, it can do more, like tell you about misassemblies, however, if the assembly is de novo then you do not have a reference.

    # Create bash script to run quast program
     vi quast.sh
    
Press [i] for insert mode and copy the below script
    
    #!/bin/bash

    #SBATCH --account iacc_jfierst
    #SBATCH --qos=highmem1
    #SBATCH --partition=highmem1
    #SBATCH --output=out_quast.log
    #SBATCH --mail-user=username@email.com   # Please use your own email
    #SBATCH --mail-type=ALL

    module load quast-5.2.0  
   
    # Replace (/your/path/to/your/genome_assembly_file.fasta) with the path to your assemble file, press clt + v (for windows), command + v (for macs) to paste the assembly file path (This is a comment please don't copy it)
    quast.py -t 4 --plots-format pdf /your/path/to/your/genome_assembly_file.fasta -o ./quast_out/
    
    # Don't forget to add your assembly file (nd.asm.fasta or assembly.fasta) after pasting the file path (This is a comment please don't copy it)
    
Save by pressing [esc], type ':w' (without the single quotes). To exit type: ':q' (without the single quotes) and press [enter]

Now let's run the bash script Type:

     sbatch quast_run.sh
    
To see if your job is running type:

      squeue --me


The html files are files that display the information in a graphical way using icarus viewer. To view these files, you need to download them to your local machine and then click on them to open them.

After the job is finished you'll see a new directory called quast_out. If you cd (change directory) into the nematode_quast dirctory you'll see a file called icarus.html. We need to download this file on to our local machine (your personal laptop) to view its contents.
 
First go to the HPC dashboard tab that is currently open on your laptop. 

Click on the Files button.
Next, click on the Home Directory button.
Navigate to the quast_out directory.

Select the quast_out directory, by clicking on the white box next to quast_out directory.

Finally, click the Download button in the top-right corner to download the directory.
(Isn't this harder than the command?)

Once you have download the file, double click the report.html file to view the assembly statistics for your assembly!!

So what's your N50 and L50?

</details>

<details>
<summary><b>BUSCO</b></summary>
 
    # Create bash script to run busco program
      vi busco.sh
Press [i] for insert mode and copy the below script

    #!/bin/bash

    #SBATCH --account iacc_jfierst
    #SBATCH --qos=highmem1
    #SBATCH --partition=highmem1
    #SBATCH --output=out_busco.log
    #SBATCH -n 12
    #SBATCH --mail-user=username@email.com   # Please use your own email
    #SBATCH --mail-type=ALL

    module load busco/5.4.7
    
    # Replace (/your/path/to/your/genome_assembly_file.fasta) with the path to your assemble file. (This is a comment please don't copy it)
    busco -c 4 -m genome -i /your/path/to/your/genome_assembly_file.fasta -o busco_out --offline --lineage_dataset /home/data/jfierst/nanopore_class_2025/assemblyPractice/bacteria_odb10

    # Don't forget to add your assembly file (nd.asm.fasta or assembly.fasta) after pasting the file path (This is a comment please don't copy it)

Save by pressing [esc], type ':w' (without the single quotes). To exit type: ':q' (without the single quotes) and press [enter]
Now let's run the bash script Type:

     sbatch busco.sh
    
To see if your job is running type:

      squeue --me
Once the job is finished you should see a new directory called busco_out, go into the dirctory with the cd command. Use the more commmand to see the file contents in short_summary.specific.bacteria_odb10.busco_out.txt
    
    more short_summary.specific.bacteria_odb10.busco_out.txt

The simple output of BUSCO is in short_summary.specific.bacteria_odb10.busco_out.txt. The line you want to pay attention to is:

    # This is the BUSCO result of the flye genome assembly
    C:100.0%[S:100.0%,D:0.0%],F:0.0%,M:0.0%,n:124
    

Here, 124 orthologous genes were searched from the bacterial dataset. All 124 were found in the genome assembly, complete and single copy. None were duplicated or fragmented. This is perfect, literally. High duplication may be a result of contamination or heterozygosity in the genome assembly. High missing or fragmentation may mean some data is missing (not enough coverage or a particularly difficult part to assemble in the genome), or if the ortholog has mutated in some way that it cannot be aligned.

</details>


## Try experimenting with the input data

    module avail seqtk

    module load seqtk-1.3 #(or whichever version you found)

    seqtk sample -s 3 ecoli.fastq .25 > subsample.fasta

This command is using seqtk to take a random sample of 25% of of the fastq reads and save them to subsample.fastq. -s is used to set the seed, allowing the 'random' sampling to be reproducable if given the same seed (3). You could change this to any number. How would the assembly turn out with 1/4th of the data?

Alternatively, what if we only used reads that were ultra long, say above 50kb?

    seqkit seq -m 50000 ecoli.fastq > 50kb_subsample.fasta

 How would this type of input data affect the assembly?
 



