## FIU Genomics & Bioinformatics Analysis CURE BSC3990L Genomics ###

***Genome Annotation tutorial***

<b>Note:</b> This tutorial demonstrates eukaryotic <i>de novo</i> genome annotation without RNA data. Although we have been using <i>E.coli</i> in our previous tutorials, it is not possible with some of the softwares used in this pipeline. Why? because some rely on databases specific to eukaryotes. So, we will be demonstrating with <i>c.elegans</i>. Because we are using <i>c.elegans</i>, this is not a quick and short tutorial. Some of these programs can take several hours.

1. Bacterial genome annotation has its own specialized softwares which are likely to perform better. If you are doing prokaryotic genome annotation, try looking into [Prokka](https://github.com/tseemann/prokka).

2. <i>De novo</i> annotation is more difficult than if you had a reference genome. For gene annotation when a reference is present, try looking into [Liftoff](https://github.com/agshumate/Liftoff).

3. The more data the better. For this course we did not do RNA sequencing, however, it is recommended to provide more evidence for annotation softwares. 
If you have RNA data, the process is very similar. The only difference is that after repeat masking, you have to align your RNA data to your genome. 
A few popular tools exist for this; try looking into [STAR](https://github.com/alexdobin/STAR?tab=readme-ov-file) or [HISAT2](https://daehwankimlab.github.io/hisat2/manual/). 
You can find more information on running STAR or HISAT2 on the FIU HPC [here](https://github.com/ToriEggers/DeNovo-Nematode-Pipeline/blob/main/DeNovo-Nematode-Pipeline.md), under protein annotation. 

<details>

<summary><b>Masking Repeats</b></summary>

Masking the repeat regions of the genome makes protein annotation easier, and many softwares ask for a 'softmasked' version of a genome. If not provided, they often make one for you before beginning annotation. 'Softmasking' a genome means to replace repeat regions with lowercase letters, so instead of ATGCGC, you might see ATgcgc. 'Hardmasking' a genome means to replace repeat regions with 'N', so instead of ATGCGC, you might see ATNNNN. Here, we softmask our genome with repeatmodeler and repeatmasker. 

## Set up

`cd` to your working directory. For me this is /home/data/jfierst_classroom/tori

```
mkdir annotation
```

```
cd annotation
```

## Get the Data

There are 2 ways to get the data (if you are not affiliated with the class do the second option):

Option #1
```
cp /home/data/jfierst_classroom/annotationPractice/elegans.fasta
```

Option #2

<b> Do not copy/paste teh entire code block but go line by line</b>
```
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/985/GCF_000002985.6_WBcel235/GCF_000002985.6_WBcel235_genomic.fna.gz
gunzip GCF_000002985.6_WBcel235_genomic.fna.gz
awk '/^>/ {print; next} {print toupper($0)}' GCF_000002985.6_WBcel235_genomic.fna > elegans.fasta
```

## Generate a Repeat Library

First we use [RepeatModeler/Masker](https://github.com/Dfam-consortium/RepeatModeler) to find repetitive regions and create whats called a 'repeat library'.

Install repeatModeler/Masker with conda:

```
module load mamba/23.1.0-4
```

```
conda create -n repeatmodeler
```

```
source activate repeatmodeler
```

```
mamba install -c bioconda repeatmodeler
```
Make sure it installed properly. Type `BuildDatabase` and the manual for that command should appear. If there is an error that says command not found, make sure you did the above installation correctly. 

Make the script:
```
vi repeatmodeler.sh
```

Hit [i] for insert and copy/paste the following:
```
#!/bin/bash

#SBATCH --account acc_jfierst_classroom
#SBATCH --qos highmem1
#SBATCH --partition highmem1
#SBATCH --output=out_repeatmodeler.log
#SBATCH --mail-user=username@email.com   #use your own email
#SBATCH --mail-type=ALL

module load mamba/23.1.0-4
source activate repeatmodeler

#Build the database
BuildDatabase -name ELEGANS elegans.fasta

#Run RepeatModeler for de novo repeat identification and characterization. Takes long time.
RepeatModeler -threads 8 -database ELEGANS

#Use the queryRepeatDatabase.pl script inside RepeatMasker/util to extract Rhabditida repeats
python /home/[username]/.conda/envs/repeatmodeler/share/RepeatMasker/famdb.py families -f fasta_acc -ad --curated 'rhabditida' > Rhabditida.repeatmasker

#Combine the files to create a library of de novo and known repeats
cat RM*/consensi.fa.classified Rhabditida.repeatmasker > Elegans.repeats

```

Hit [Esc], then type `:wq` and hit [Enter] 

Your output will be Elegans.repeats. This takes about 1 day to complete on <i>c.elegans</i>


</details>


<details>
<summary><b>Protein Prediction</b></summary>

Many softwares exist for protein prediction, most of which use machine learning methods like CNNs, LSTMs, and HMMs. Those used today include BRAKER, GALBA, Tiberius, and Helixer. Some older ones include MAKER or FunAnnotate. BRAKER has 3 versions: BRAKER1 is genome + RNA, BRAKER2 is genome + protein, and BRAKER3 is genome + RNA + protien. GALBA is like BRAKER2 but specifically code to deal with issues unique to large genomes (>500Gb). Tiberius is the newest annotator but has only been trained (modeled) on mammalian data. Helixer has more available models but is more particular about GPU requirements. Below, we use BRAKER2 for gene annotation, although notice from the images that these may not be the best annotations. Rather, it's what we can do with what we have.

![screenshot](https://github.com/FierstLab/Bootcamp/blob/main/pictures/annotatorsF1.png)

![screenshot](https://github.com/FierstLab/Bootcamp/blob/main/pictures/annotatorsRecall.png)

## BRAKER



## AGAT



</details>



<details>
<summary><b>Functional Annotation</b></summary>

 ## InterProScan


</details>


<details>

<summary><b>Repeat Annotation</b></summary>

There are many softwares for repeat annotation. Extensive de novo repeat annotator (EDTA) was made by plant biologists and is sort of the standard in the field at the moment. However, many other softwares have been released in recent years, such as EarlGrey (benchmarked with <i>D.melanogaster</i>) and TransposonUltimate (benchmarked with <i>C.elegans</i>). All of these programs have the same outputs: (1) a repeat library in fasta format (2) a fasta of all repeats in the genome and (3) a gff file of all repeats in the genome.

## EDTA

## EarlGrey

## TransposonUltimate

</details>


